ESLint cannot find errors connected with react prop-types when using mobX decorators with stateless components

<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).

    This template is for bug reports. If you are here for another reason, please see below:

    1. To propose a new rule: https://eslint.org/docs/developer-guide/contributing/new-rules
    2. To request a rule change: https://eslint.org/docs/developer-guide/contributing/rule-changes
    3. To request a change that is not a bug fix, rule change, or new rule: https://eslint.org/docs/developer-guide/contributing/changes
    4. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

    Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

* **ESLint Version: 5.16.0**
* **Node Version: 10.15.0**
* **npm Version: 6.4.1**

**What parser (default, Babel-ESLint, etc.) are you using?**

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

 package.json 
```
{
  "name": "",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "start": "node node_modules/react-native/local-cli/cli.js start",
    "test": "jest",
    "lint": "eslint **/*.js",
    "precommit": "lint-staged",
    "android": "cd ./android && ./gradlew app:assembleDebug && ./gradlew installDebug"
  },
  "rnpm": {
    "assets": [
      "src/resources/fonts"
    ]
  },
  "dependencies": {
    "@react-native-community/async-storage": "^1.3.4",
    "axios": "^0.18.0",
    "axios-mock-adapter": "^1.16.0",
    "final-form": "^4.12.0",
    "mobx": "^4.9.4",
    "mobx-react": "^5.4.3",
    "prop-types": "^15.7.2",
    "react": "16.8.3",
    "react-final-form": "^4.1.0",
    "react-native": "0.59.5",
    "react-native-config": "^0.11.7",
    "react-native-navigation": "^2.18.4",
    "react-native-paper": "^2.15.2",
    "react-native-vector-icons": "^6.4.2",
    "styled-components": "^4.2.0"
  },
  "devDependencies": {
    "@babel/core": "^7.4.4",
    "@babel/plugin-proposal-class-properties": "^7.4.4",
    "@babel/plugin-proposal-decorators": "^7.4.4",
    "@babel/runtime": "^7.4.4",
    "babel-eslint": "^10.0.1",
    "babel-jest": "^24.7.1",
    "enzyme": "^3.9.0",
    "enzyme-adapter-react-16": "^1.12.1",
    "eslint": "^5.16.0",
    "eslint-config-airbnb": "^17.1.0",
    "eslint-plugin-import": "^2.17.2",
    "eslint-plugin-jsx-a11y": "^6.2.1",
    "eslint-plugin-react": "^7.13.0",
    "eslint-plugin-react-native": "^3.7.0",
    "jest": "^24.7.1",
    "metro-react-native-babel-preset": "^0.54.0",
    "mobx-remotedev": "^0.3.6",
    "react-dom": "^16.8.6",
    "react-test-renderer": "16.8.3"
  },
  "jest": {
    "preset": "react-native"
  }
}
```

<!-- Paste your configuration below: -->
.eslintrc.json
```js
{
    "extends": "airbnb",
    "parser": "babel-eslint",
    "env": {
      "browser": true,
      "jest": true
    },
    "plugins": [
      "import",
      "jsx-a11y",
      "react",
      "react-native"
    ],
    "rules": {
      "import/extensions": "off",
      "import/no-extraneous-dependencies": "off",
      "import/no-unresolved": "off",
      "linebreak-style": "off"
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

<!-- Paste the source code below: -->
```js
const SideMenu = observer(
  ({ componentId }) => (
    <S.Container>
      <S.Head />
      <UserInfo />
      <Separator />
      <MenuList componentId={componentId} />
    </S.Container>
  ),
);

```

<!-- Paste the command you used to run ESLint: -->
```bash
eslint **/*.js
```

**What did you expect to happen?** - That `` componentId `` is missing in props validation


**What actually happened? Please include the actual, raw output from ESLint.** Nothing happens. Сode passes verification successfully
```
yarn run v1.12.3
$ eslint **/*.js

<rootFolder>\src\modules\Main\UserInfo\Form\index.js
  10:28  warning  Unexpected console statement  no-console

✖ 1 problem (0 errors, 1 warning)

Done in 9.38s.
```


**Are you willing to submit a pull request to fix this bug?** - No

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,11 @@ This change log adheres to standards from [Keep a CHANGELOG](http://keepachangel
 
 ## Unreleased
 
+### Added
+* component detection: add componentWrapperFunctions setting ([#2713][] @@jzabala @LandonSchropp)
+
+[#2713]: https://github.com/yannickcr/eslint-plugin-react/pull/2713
+
 ## [7.23.2] - 2021.04.08
 
 ### Fixed

--- a/README.md
+++ b/README.md
@@ -52,6 +52,13 @@ You should also specify settings that will be shared across all the plugin rules
         {"property": "freeze", "object": "Object"},
         {"property": "myFavoriteWrapper"}
     ],
+    "componentWrapperFunctions": [
+        // The name of any function used to wrap components, e.g. Mobx `observer` function. If this isn't set, components wrapped by these functions will be skipped.
+        "observer", // `property`
+        {"property": "styled"} // `object` is optional
+        {"property": "observer", "object": "Mobx"},
+        {"property": "observer", "object": "<pragma>"}, // sets `object` to whatever value `settings.react.pragma` is set to
+    ],
     "linkComponents": [
       // Components used as alternatives to <a> for linking, eg. <Link to={ url } />
       "Hyperlink",

--- a/lib/util/Components.js
+++ b/lib/util/Components.js
@@ -212,11 +212,28 @@ class Components {
   }
 }
 
+function getWrapperFunctions(context, pragma) {
+  const componentWrapperFunctions = context.settings.componentWrapperFunctions || [];
+
+  // eslint-disable-next-line arrow-body-style
+  return componentWrapperFunctions.map((wrapperFunction) => {
+    return typeof wrapperFunction === 'string'
+      ? {property: wrapperFunction}
+      : Object.assign({}, wrapperFunction, {
+        object: wrapperFunction.object === '<pragma>' ? pragma : wrapperFunction.object
+      });
+  }).concat([
+    {property: 'forwardRef', object: pragma},
+    {property: 'memo', object: pragma}
+  ]);
+}
+
 function componentRule(rule, context) {
   const createClass = pragmaUtil.getCreateClassFromContext(context);
   const pragma = pragmaUtil.getFromContext(context);
   const sourceCode = context.getSourceCode();
   const components = new Components();
+  const wrapperFunctions = getWrapperFunctions(context, pragma);
 
   // Utilities for component detection
   const utils = {
@@ -597,14 +614,20 @@ function componentRule(rule, context) {
       if (!node || node.type !== 'CallExpression') {
         return false;
       }
-      const propertyNames = ['forwardRef', 'memo'];
-      const calleeObject = node.callee.object;
-      if (calleeObject && node.callee.property) {
-        return arrayIncludes(propertyNames, node.callee.property.name)
-          && calleeObject.name === pragma
-          && !this.nodeWrapsComponent(node);
-      }
-      return arrayIncludes(propertyNames, node.callee.name) && this.isDestructuredFromPragmaImport(node.callee.name);
+
+      return wrapperFunctions.some((wrapperFunction) => {
+        if (node.callee.type === 'MemberExpression') {
+          return wrapperFunction.object
+            && wrapperFunction.object === node.callee.object.name
+            && wrapperFunction.property === node.callee.property.name
+            && !this.nodeWrapsComponent(node);
+        }
+        return wrapperFunction.property === node.callee.name
+          && (!wrapperFunction.object
+            // Functions coming from the current pragma need special handling
+            || (wrapperFunction.object === pragma && this.isDestructuredFromPragmaImport(node.callee.name))
+          );
+      });
     },
 
     /**

--- a/tests/lib/rules/prop-types.js
+++ b/tests/lib/rules/prop-types.js
@@ -2515,6 +2515,26 @@ ruleTester.run('prop-types', rule, {
         }
       `
     },
+    {
+      code: `
+      const SideMenu = styled(
+        ({ componentId }) => (
+          <S.Container>
+            <S.Head />
+            <UserInfo />
+            <Separator />
+            <MenuList componentId={componentId} />
+          </S.Container>
+        ),
+      );
+      SideMenu.propTypes = {
+        componentId: PropTypes.string.isRequired
+      }
+      `,
+      settings: {
+        componentWrapperFunctions: [{property: 'styled'}]
+      }
+    },
     parsers.TS([
       {
         code: `
@@ -6035,6 +6055,45 @@ ruleTester.run('prop-types', rule, {
       errors: [{
         messageId: 'missingPropType',
         data: {name: 'name'}
+      }]
+    },
+    {
+      code: `
+      const SideMenu = observer(
+        ({ componentId }) => (
+          <S.Container>
+            <S.Head />
+            <UserInfo />
+            <Separator />
+            <MenuList componentId={componentId} />
+          </S.Container>
+        ),
+      );`,
+      settings: {
+        componentWrapperFunctions: ['observer']
+      },
+      errors: [{
+        message: '\'componentId\' is missing in props validation'
+      }]
+    },
+    {
+      code: `
+      const SideMenu = Mobx.observer(
+        ({ id }) => (
+          <S.Container>
+            <S.Head />
+            <UserInfo />
+            <Separator />
+            <MenuList componentId={id} />
+          </S.Container>
+        ),
+      );
+      `,
+      settings: {
+        componentWrapperFunctions: [{property: 'observer', object: 'Mobx'}]
+      },
+      errors: [{
+        message: '\'id\' is missing in props validation'
       }]
     },
     parsers.TS([

